include/oneapi/dpl/internal/gcd_impl.h

```diff
@@ -0,0 +1,108 @@
+/*
+ *  Copyright (c) Intel Corporation
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#ifndef _ONEDPL_GCD
+#define _ONEDPL_GCD
+
+#include <limits>
+#include "gcd_defs.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+
+namespace internal
+{
+template <typename _Result, typename _Source, bool _IsSigned = ::std::is_signed<_Source>::value>
```

Comment:
I would express ``__ct_abs`` via functions. For example:

```cpp
template <typename _Result, typename _Source>
_Result __ct_abs_impl(_Source __t, ::std::true_type)
{
    if (__t >= 0)
        return __t;
    if (__t == ::std::numeric_limits<_Source>::min())
        return -static_cast<_Result>(__t);
    return -__t;
};

template <typename _Result, typename _Source>
_Result __ct_abs_impl(_Source __t, ::std::false_type)
{
    return __t;
};

template <typename _Result, typename _Source>
constexpr _Result __ct_abs(_Source __t) {
    return __ct_abs_impl<_Result>(__t, ::std::is_signed<_Source>{});
}

```

I also don't know what is `__ct`. Is it _class template_? Anyway I would give to it more meaningful name like __calc_abs or __get_abs or something like that.

The call of the functions would also be simplified: like ``oneapi::dpl::internal::__ct_abs<_Rp>(__m)``

Reply:
Full implementation is here: https://godbolt.org/z/aTnn71

Reply:
Thanks for your comments. Done.